docs/DafnyRef/Expressions.md

```diff
@@ -913,7 +913,24 @@ var k := 10 / x; // error, may divide by 0.
 var m := if x != 0 then 10 / x else 1; // ok, guarded
 ```
 
-TO BE WRITTEN - binding form
+The `if` expression also permits a binding form.
+In this case the condition of the if is an existential asking
+"does there exist a value satisfying a given predicate?".
+If not, the else branch is evaluated. But if so, then an
+(arbitrary) value that does satisfy the given predicate is
+bound to the given variable and that variable is in scope in 
+the then-branch of the expression.
+
+For example, in the expression
+```
+if x: int :| x*x == s then x else 0
+```
+if `s` is indeed a square, say `25`, then x is bound to either `5` or `-5`
```

Comment:
From an engineer's perspective where everything is deterministic, this is weird. My interpretation is that the value is either 5 or -5 in ghost code so that it's just a predicate, but what happens in non-ghost code? Can you write an explanation?

Also, could you perhaps use another example than something non-linear? Non-linearity give headaches to z3 in general.
For example, I suggest:

```dafny
predicate P(x: int) {
  x == 5 || x == -5
}
method main() {
  assert P(5);
  var y := if x: int :| P(x) then x else 0;
  assert y == 5 || y == -5;
}
```

Here if you change `x: int` into `x: nat`, you can remove the `y == -5`;
In the case above, if I add a range, then the `if` expression is no longer ghost:
```dafny
 var y := if x: nat :| 0 <= x < 10 && P(x) then x else 0;
```
Also, since the expression hides an existential quantifier, one needs to write a trigger to prove anything concerning its value, that's why I wrote `assert P(5);` above, without which it fails.

Reply:
Used your example and reworded.